nixos/modules/config/mysql.nix

```diff
@@ -0,0 +1,485 @@
+{ config, pkgs, lib, ... }:
+
+with lib;
+
+let
+  cfg = config.users.mysql;
+in
+{
+  options = {
+    users.mysql = {
+      enable = mkEnableOption "Authentication against a MySQL/MariaDB database";
+      host = mkOption {
+        type = types.str;
+        example = "localhost";
+        description = "The hostname of the MySQL/MariaDB server";
+      };
+      database = mkOption {
+        type = types.str;
+        example = "auth";
+        description = "The name of the database containing the users";
+      };
+      user = mkOption {
+        type = types.str;
+        example = "nss-user";
+        description = "The username to use when connecting to the database";
+      };
+      passwordFile = mkOption {
+        type = types.path;
+        example = "/etc/nss-db-passwd";
+        description = "The path to the file containing the password for the user";
+      };
+      pam = mkOption {
+        description = "Settings for <literal>pam_mysql</literal>";
+        type = types.submodule {
+          options = {
+            table = mkOption {
+              type = types.str;
+              example = "users";
+              description = "The name of table that maps unique login names to the passwords.";
+            };
+            updateTable = mkOption {
+              type = types.nullOr types.str;
+              default = null;
+              example = "users_updates";
+              description = ''
+                The name of the table used for password alteration. If not defined, the value
+                of the <literal>table</literal> option will be used instead.
+              '';
+            };
+            userColumn = mkOption {
+              type = types.str;
+              example = "username";
+              description = "The name of the column that contains a unix login name.";
+            };
+            passwordColumn = mkOption {
+              type = types.str;
```

Comment:
That should be possible to supply via a file.

Reply:
This is only the column in the MySQL database which contains the user passwords. I don't think that this is so critical that it has to be supplied via a file.

Reply:
secrets like passwords should never be supplied via string because then they end up in plain text in the nix store which is world readable.

Reply:
Again: that is not a password. It is just the name of the column in the users table of the database which _stores_ the passwords.

So the content of this option is just a column name and this is not critical in my opinion.

Reply:
I try to explain it with an example:

Imagine that the setup is the following:

The authentication data is stored in a MySQL database on the host `mysql.example.com` in the database `auth`, which can be accessed by the MySQL-User `nixos_auth` and the MySQL-Password `example123`.

In the `auth` database is a table called `users` with the columns `id`, `username`, `hashedPassword` (there are still missing columns for `gids` and so for `libnss-mysql`, but this is just a minimal example for the `passwordColumn` of `pam_mysql`).

This would result in the following (incomplete) `users.mysql` configuration:

```nix
host = "mysql.example.com";
database = "auth";
user = "nixos_auth";
# /run/secrets/mysql-auth-pass contains "example123"
passwordFile = "/run/secrets/mysql-auth-pass";
pam.usernameColumn = "username";
pam.passwordColumn = "hashedPassword";
```

So `pam.passwordColumn` is literally just the column in the SQL table which contains the (hashed) passwords. A attacker can do nothing with just the column name, as long as he cannot connect to the database using the password from the password file `passwordFile`.

And if the attacker knows the password from password file, he can just run a `SELECT * FROM users` and then he has the complete content of the table. So he does not need the password column name at all.

So there is no reason to store the name of the `passwordColumn` in a file, because it is just the name of an SQL table column and nothing that is used to authenticate against something.

Reply:
I think it was a misunderstanding, as @netalidev stated, no secret is here -> resolving.